momento-provider/src/main/java/com/google/code/ssm/providers/momento/MomentoClientWrapper.java

```diff
@@ -217,11 +196,67 @@ public Object getNativeClient() {
         return transcoderAdapter;
     }
 
-    private static int getFlag(byte[] data, int i) {
-        return (data[i] & 0xff) << 24
-                | (data[i + 1] & 0xff) << 16
-                | (data[i + 2] & 0xff) << 8
-                | (data[i + 3] & 0xff);
+    // Helper function that concats the CachedObject to raw bytes before calling `set`.
+    // This helps maintain our serialization flags so that we can deserialize the object upon a `get`
+    private boolean writeOutToMomento(final String key, final int exp, final int flagsUsed, final byte[] data) {
+        final byte[] flags = intToBytes(flagsUsed);
+        ByteBuffer buffer = ByteBuffer.allocate(flags.length + data.length);
+        buffer.put(flags);
+        buffer.put(data);
+        buffer.rewind();
+        CacheSetResponse response = momentoClient.set(defaultCacheName, key, buffer.asReadOnlyBuffer(), exp);
+        return response != null;
+    }
+
+    // Helper function that reads the raw bytes from Momento and splits up the bytes
+    // to extract the serialization flags we used as well as the raw bytes of data that contain our
+    // actual object
+    private <T> T readFromMomento(final String key, final CacheTranscoder transcoder) {
+        Transcoder<T> cacheTranscoder = getTranscoder(transcoder);
+        CacheGetResponse response = momentoClient.get(defaultCacheName, key);
+        Optional<byte[]> cacheGetResponse = response.byteArray();
+        if (cacheGetResponse.isPresent()) {
+            byte[] returnedBytes = cacheGetResponse.get();
+            int flags = getFlags(returnedBytes);
+            byte[] originalData = getData(returnedBytes);
+            return cacheTranscoder.decode(new CachedData(flags, originalData, originalData.length));
+        }
+        return null;
+    }
+
+    // Helper function that reads the raw bytes from Momento and splits up the bytes
+    // to extract the serialization flags we used as well as the raw bytes of data that contain our
+    // actual object, uses our default Transcoder
+    private Object readFromMomento(final String key) {
+        CacheTranscoder cacheTranscoder = getTranscoder();
+        CacheGetResponse response = momentoClient.get(defaultCacheName, key);
+        Optional<byte[]> cacheGetResponse = response.byteArray();
+        if (cacheGetResponse.isPresent()) {
+            byte[] returnedBytes = cacheGetResponse.get();
+            int flags = getFlags(returnedBytes);
+            byte[] originalData = getData(returnedBytes);
+            return cacheTranscoder.decode(new CachedObjectWrapper(new CachedData(flags, originalData, originalData.length)));
+        }
+        return null;
+    }
+
+    private static byte[] intToBytes(final int i) {
+        ByteBuffer byteBuffer = ByteBuffer.allocate(EXTRA_HDR_LEN);
+        byteBuffer.putInt(i);
+        return byteBuffer.array();
+    }
```

Comment:
`writeOutToMomento` uses `ByteBuffer`. This method does an unnecessary 4 bytes' heap array allocation.

I think you want to delete this method and replace its 1 usage on line 204 with `buffer.putInt(flagsUsed);`

Reply:
Sounds good to me!